internal/setup/setup.go

```diff
@@ -107,309 +96,51 @@ func StartSetup(esConfig es_int.Config, sd systemdefaults.SystemDefaults) (*Setu
-	if err != nil {
-		logging.Log("SETUP-0874m").WithError(err).Error("unable to set global org")
-		return err
-	}
 
-	err = setUp.setIamProject(ctx, setUpConfig.IAMProject)
-	if err != nil {
-		logging.Log("SETUP-kaWjq").WithError(err).Error("unable to set zitadel project")
-		return err
-	}
-
-	iam, err = s.IamEvents.SetupDone(ctx, s.iamID)
-	if err != nil {
-		logging.Log("SETUP-de342").WithError(err).Error("unable to finish setup")
-		return err
-	}
-	logging.Log("SETUP-ds31h").Info("setup done")
-	return nil
-}
-
-func (setUp *initializer) loginPolicy(ctx context.Context, policy LoginPolicy) error {
-	logging.Log("SETUP-4djul").Info("setting up login policy")
-	loginPolicy := &iam_model.LoginPolicy{
-		ObjectRoot: models.ObjectRoot{
-			AggregateID: setUp.iamID,
-		},
-		AllowRegister:         policy.AllowRegister,
-		AllowUsernamePassword: policy.AllowUsernamePassword,
-		AllowExternalIdp:      policy.AllowExternalIdp,
-	}
-	_, err := setUp.IamEvents.AddLoginPolicy(ctx, loginPolicy)
-	return err
-}
-
-func (setUp *initializer) orgs(ctx context.Context, orgs []Org) error {
-	logging.Log("SETUP-dsTh3").Info("setting up orgs")
-	for _, iamOrg := range orgs {
-		org, err := setUp.org(ctx, iamOrg)
-		if err != nil {
-			logging.LogWithFields("SETUP-IlLif", "Org", iamOrg.Name).WithError(err).Error("unable to create org")
-			return err
-		}
-		setUp.createdOrgs[iamOrg.Name] = org
-
-		var policy *org_model.OrgIAMPolicy
-		if iamOrg.OrgIamPolicy {
-			policy, err = setUp.iamorgpolicy(ctx, org)
-			if err != nil {
-				logging.LogWithFields("SETUP-IlLif", "Org IAM Policy", iamOrg.Name).WithError(err).Error("unable to create iam org policy")
-				return err
-			}
-		} else {
-			policy, err = setUp.OrgEvents.GetOrgIAMPolicy(ctx, policy_model.DefaultPolicy)
-			if err != nil {
-				logging.LogWithFields("SETUP-IS8wS", "Org IAM Policy", iamOrg.Name).WithError(err).Error("unable to get default iam org policy")
-				return err
-			}
+	for _, step := range steps {
+		step.init(s)
+		if step.step() != iam.SetUpDone+1 {
+			logging.LogWithFields("SETUP-rxRM1", "step", step.step(), "previous", iam.SetUpDone).Warn("wrong step order")
+			return errors.ThrowPreconditionFailed(nil, "SETUP-wwAqO", "too few steps for this zitadel verison")
 		}
-
-		ctx = setSetUpContextData(ctx, org.AggregateID)
-		err = setUp.users(ctx, iamOrg.Users, policy)
+		iam, err = s.IamEvents.StartSetup(ctx, s.iamID, step.step())
 		if err != nil {
-			logging.LogWithFields("SETUP-8zfwz", "Org", iamOrg.Name).WithError(err).Error("unable to set up org users")
 			return err
 		}
 
-		err = setUp.orgOwners(ctx, org, iamOrg.Owners)
+		err = step.execute(ctx)
 		if err != nil {
-			logging.LogWithFields("SETUP-0874m", "Org", iamOrg.Name).WithError(err).Error("unable to set up org owners")
 			return err
 		}
 
-		err = setUp.projects(ctx, iamOrg.Projects, setUp.createdUsers[iamOrg.Owners[0]].AggregateID)
+		iam, err = s.IamEvents.SetupDone(ctx, s.iamID, step.step())
```

Comment:
set up done should be in execute, to have it in one transcation.
Otherwise it could be, that everything from the step is done, but the done event is missing

Reply:
In my opinion it would make more sense to implement that change after [this one](https://github.com/caos/zitadel/issues/680) because it will have less impact to the code than now. We have the problem with the single transaction since we implemented the setup functionality.